hadoop-hdds/docs/content/design/leader-execution/obs-locking.md

```diff
@@ -0,0 +1,97 @@
+| CommitKey               | Bucket Read, Key Write lock               | Avoid parallel key commit by different client, otherwise it can leave dangling blocks on overwrite          |
+| InitiateMultiPartUpload | Bucket Read Lock, `No Lock` for key       | Key can be created parallel by client with different uploadId in open key table, so do not need key lock    |
+| CommitMultiPartUpload   | WriteLock: PartKey Name                   | Avoid same part commit in parallel, else it can leave dangling blocks on overwrite                          |
+| CompleteMultiPartUpload | Bucket Read, Key Write lock               | Avoid parallel multi-part upload complete with different uploadId, else overwrite can cause dangling blocks |
+| AbortMultiPartUpload    | Bucket Read, Key Write lock               | Need to avoid other operation in parallel like commit part                                                  |
+| DeleteKey               | Bucket Read, Key Write lock               | Avoid create and delete in parallel                                                                         |
+| DeleteKeys              | Bucket Read, Key Write with ordered lock  | Avoid create and delete in parallel, ordered key locks to avoid deadlock                                    |
+| RenameKey               | Bucket Read, Keys Write with ordered lock | lock in order to delete key from original location and move to new location                                 |
+| SetAcl                  | Bucket Read, Key Write lock               | Avoid parallel delete key                                                                                   |
+| AddAcl                  | Bucket Read, Key Write lock               | Avoid parallel delete key                                                                                   |
+| RemoveAcl               | Bucket Read, Key Write lock               | Avoid parallel delete key                                                                                   |
+| AllocateBlock           | Bucket Read, Key Write lock               | Need lock key to avoid wrong update of key, if same client does parallel allocate                           |
+| SetTimes                | Bucket Read, Key Write lock               | Avoid parallel delete key                                                                                   |
+
+Batch Operation:
+1. deleteKeys: batch will be divided to multiple threads in Execution Pool to run parallel calling DeleteKey
+2. RenameKeys: This is `depreciated`, but for compatibility, will be divided to multiple threads in Execution Pool to run parallel calling RenameKey
+
+For batch operation, atomicity is not guranteed for above api, and same is behavior for s3 perspective.
+
+## Bucket and volume locking as required for concurrency for obs key handling
+
+### Volume Operation
```

Comment:
We probably should start with Volume, then Bucket and then Key.

Reply:
done